cmd/up.go

```diff
@@ -42,8 +52,197 @@ func newUpCmd() *cobra.Command {
+		}
+
+		// Retrieve the template repo.
+		repo, err := workspace.RetrieveTemplates(url, false)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			contract.IgnoreError(repo.Delete())
+		}()
+
+		// List the templates from the repo.
+		templates, err := repo.Templates()
+		if err != nil {
+			return err
+		}
+
+		// Make sure only a single template is found.
+		// Alternatively, we could consider prompting to choose one instead of failing.
+		if len(templates) != 1 {
+			return errors.Errorf("more than one application found at %s", url)
+		}
+		template := templates[0]
+
+		// Create temp directory for the "virtual workspace".
+		temp, err := ioutil.TempDir("", "pulumi-up-")
+		if err != nil {
+			return err
+		}
+		defer func() {
+			contract.IgnoreError(os.RemoveAll(temp))
+		}()
+
+		// Cleanup the project name/description if needed.
+		projectName := template.ProjectName
+		if projectName == "${PROJECT}" {
+			projectName = workspace.ValueOrSanitizedDefaultProjectName(projectName, template.Name)
+		}
+		projectDescription := template.ProjectDescription
+		if projectDescription == "${DESCRIPTION}" {
+			projectDescription = ""
+		}
+
+		// Copy the template files from the repo to the temporary "virtual workspace" directory.
+		if err = template.CopyTemplateFiles(temp, true, projectName, projectDescription); err != nil {
+			return err
+		}
+
+		// Change the working directory to the "virtual workspace" directory.
+		if err = os.Chdir(temp); err != nil {
+			return errors.Wrap(err, "changing the working directory")
+		}
+
+		// Get a stack, but don't set it to current.
+		s, err := requireStack(stack, true, opts.Display, false /*setCurrent*/)
+		if err != nil {
+			return err
+		}
+
+		// Get the existing config. stackConfig will be nil if there wasn't a previous deployment.
+		stackConfig, err := backend.GetLatestConfiguration(commandContext(), s)
+		if err != nil && err != backend.ErrNoPreviousDeployment {
+			return err
+		}
+
+		// Get the existing snapshot.
+		snap, err := s.Snapshot(commandContext())
+		if err != nil {
+			return err
+		}
+
+		// Handle config.
+		// If this is an initial preconfigured empty stack (i.e. configured in the Pulumi Console),
+		// use its config without prompting.
+		// Otherwise, use the values specified on the command line and prompt for new values.
+		// If the stack already existed and had previous config, those values will be used as the defaults.
+		var c config.Map
+		templateURL := getPreconfiguredEmptyStackTemplateURL(stackConfig, snap)
+		if templateURL != "" {
+			c = stackConfig
+			// TODO consider warning if the specific URL is different from templateURL.
+		} else {
+			// Get config values passed on the command line.
+			commandLineConfig, err := parseConfig(configArray)
+			if err != nil {
+				return err
+			}
+
+			// Get the crypter if needed.
+			var crypter config.Crypter
+			if stackConfig != nil && stackConfig.HasSecureValue() {
```

Comment:
Another bug ☹️. Need to always pass a `Crypter`, not just when there's existing stack config that has a secure value. Otherwise, if the template has a secret, we'll try to encrypt after prompting, but panic because of nil pointer. (Or likely better, pass the stack to `promptForConfig` and have it get the `Crypter`, which would take care of it for `new` as well).